### PR TITLE
Branches: switch doesn't rely on old branch to exist

### LINKED ISSUE #1241: Unable to switch branches with `edgedb branch switch` after force dropping the current branch
Trying to switch branches after force-dropping the current branch produces an error:
```
[2024-03-08T17:20:52Z WARN  edgedb_tokio::raw::connection] Error received from server: database <branch-name> does not exist. Severity: Error. Code: 0x4030005
edgedb error: ClientConnectionFailedError: end of stream while reading message
```

This error is also produced when running other branch commands like `list`.

Workaround: Switch branches while passing a still existing branch for -b. `edgedb -b edgedb branch switch edgedb`

- EdgeDB Version: 5.0-beta.1+8477d12
- EdgeDB CLI Version: 4.2.1-dev.1092+b2a234f
- OS Version: macOS 14.0

Steps to Reproduce:

1. Create a branch
2. Switch to the branch (done as part of create in current nightly)
3. Get a REPL to a different branch, for instance with `edgedb -b edgedb`
4. Drop the new/current branch in that REPL with `drop branch <branch-name> force;`
5. Try switching the current branch with `edgedb branch switch edgedb`

### PR DESCRIPTION
**Pull request by quinchs**
Fixes #1241

### CODE REVIEW

File: src/branch/connections.rs [hunk 1]
```diff
@@ -32,6 +33,23 @@ impl BranchConnection<'_> {
     }
 }
 
+pub async fn try_connect(connector: &Connector) -> anyhow::Result<Option<Connection>> {
```
Comment (aljazerzen) [inline src/branch/connections.rs:36]: Please document what does `Ok(None)` response here mean. That the database does not exists?

If this is the case, I think the naming convention would be `connect_if_branch_exists`. `try_connect` is used for stuff that returns a `Result` (i.e. `into()` vs `try_into()`).

File: src/branch/switch.rs [hunk 2]
```diff
@@ -1,39 +1,56 @@
...
+            }
         }
+    } else {
+        // try to connect to the target branch
+        let target_branch_connector = connector.database(&options.branch)?;
+        match try_connect(&target_branch_connector).await? {
+            Some(mut connection) => {
+                verify_server_can_use_branches(&mut connection).await?;
+            },
+            None => anyhow::bail!("The target branch doesn't exist, and cannot be created because the current branch cannot be connected to.")
```
Comment (aljazerzen) [inline src/branch/switch.rs:44]: > cannot be created

If we are not trying to create the branch, we should just print:

> "The target branch doesn't exist"

File: src/branch/switch.rs [hunk 3]
```diff
@@ -1,39 +1,56 @@
...
     eprintln!(
         "Switching from '{}' to '{}'",
         context.branch, options.branch
     );
+
     context.update_branch(&options.branch).await?;
 
     Ok(())
-}
+}
```
Comment (aljazerzen) [inline src/branch/switch.rs:56]: newline

**Review by aljazerzen — APPROVED** [review]
Looks good.

**Review by fantix — APPROVED** [review]

**Pull request merged**